Modify rules S5275,S3630: Briefly mention problems with implicit lifetime (CPP-5022) (#3824)
SonarSource/rspec: rules/S5275/cfamily/rule.adoc

== Why is this an issue?

> Because ``++reinterpret_cast++`` ignores the type system, it can perform dangerous conversions between unrelated types, leading to undefined behavior.

Reviewer: I think it is slightly unclear in the new wording what lead to UB. It could be "reinterpret_cast" (in which case the UB does not always happen) or it could be the "dangerous conversions" (in which case the wording is right)
If you don't like: "which could lead", what about:
```suggestion
Because ``++reinterpret_cast++`` ignores the type system, it can perform dangerous conversions between unrelated types that are undefined behavior.
```
Or, as in the previous rule:
```suggestion
Because ``++reinterpret_cast++`` ignores the type system, it can perform dangerous conversions between unrelated types, often leading to undefined behavior.
```

This rule reports an issue for these problematic uses of ``++reinterpret_cast++``:

* When it is used to make the compiler believe that an object in memory has a different type from its actual type because this has undefined behavior, such as casting a ``++long*++`` to ``++double*++``, even if their sizes in memory are the same (``++sizeof(long) == sizeof(double)++``).

* When it is used to cast between different levels of a complex inheritance hierarchy. A ``++static_cast++`` would apply pointer offsets to account for multiple inheritances, but ``++reinterpret_cast++`` does not.

> * When the cast breaks the rules for implicit object lifetimes since this has undefined behavior. This happens when casting to a type, for example, with a user-provided destructor or no trivial default constructor because ``++reinterpret_cast++`` does not execute any code.

Reviewer: I think the text is hard to read because it mixes a description of the detected pattern and a justification of why this pattern is detected.
```
This rule reports an issue for these problematic uses of ``++reinterpret_cast++``, which are all undefined behavior:

* When it is used to make the compiler believe that an object in memory has a different type from its actual type, such as casting a ``++long*++`` to ``++double*++``, even if their sizes in memory are the same (``++sizeof(long) == sizeof(double)++``).
* When it is used to cast between different levels of a complex inheritance hierarchy. A ``++static_cast++`` would apply pointer offsets to account for multiple inheritances, but ``++reinterpret_cast++`` does not.
* When the cast breaks the rules for implicit object lifetimes. For example,  this happens when casting to a type with a user-provided destructor or no trivial default constructor because ``++reinterpret_cast++`` does not execute any code.
```

Author: Thanks, I like your suggestion very much!

Author: Done in f3076cc361
